pkg/workspace/data_values_pre_processing.go

```diff
@@ -37,46 +38,89 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
 }
 
 func (o DataValuesPreProcessing) apply(files []*FileInLibrary) (*DataValues, []*DataValues, error) {
-	values := o.loader.schema.AsDataValue()
-	var libraryValues []*DataValues
-	for _, fileInLib := range files {
-		valuesDocs, err := o.templateFile(fileInLib)
-		if err != nil {
-			return nil, nil, fmt.Errorf("Templating file '%s': %s", fileInLib.File.RelativePath(), err)
-		}
+	_, checkSchema := o.loader.schema.(*schema.DocumentSchema)
 
-		for _, valuesDoc := range valuesDocs {
-			dv, err := NewDataValues(valuesDoc)
+	allDvs, err := o.collectDataValuesDocs(files)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// merge all Data Values YAML documents into one
+	var dvsForOtherLibraries []*DataValues
+	var dataValuesDoc *yamlmeta.Document
+	for _, dv := range allDvs {
+		switch {
+		case dv.HasLibRef():
+			dvsForOtherLibraries = append(dvsForOtherLibraries, dv)
+		case dataValuesDoc == nil:
+			err := o.loader.schema.ValidateWithValues(1)
 			if err != nil {
 				return nil, nil, err
 			}
 
-			switch {
-			case dv.HasLib():
-				libraryValues = append(libraryValues, dv)
-			case values == nil:
-				values = valuesDoc
-			default:
-				var err error
-				values, err = o.overlay(values, dv.Doc)
-				if err != nil {
-					return nil, nil, err
-				}
+			dataValuesDoc = dv.Doc
+		default:
+			if checkSchema {
+				o.annotateWithDefaultMissingOk(dv.Doc)
+			}
+			dataValuesDoc, err = o.overlay(dataValuesDoc, dv.Doc)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+		if checkSchema {
+			typeCheck := o.typeAndCheck(dataValuesDoc)
+			if len(typeCheck.Violations) > 0 {
+				return nil, nil, typeCheck
 			}
 		}
 	}
 
-	values, err := o.overlayValuesOverlays(values)
+	if dataValuesDoc == nil {
+		dataValuesDoc = o.NewEmptyDataValuesDocument()
+	}
+	dataValues, err := NewDataValues(dataValuesDoc)
 	if err != nil {
 		return nil, nil, err
 	}
+	return dataValues, dvsForOtherLibraries, nil
+}
 
-	dv, err := NewDataValues(values)
-	if err != nil {
-		return nil, nil, err
+func (o DataValuesPreProcessing) collectDataValuesDocs(files []*FileInLibrary) ([]*DataValues, error) {
+	var allDvs []*DataValues
+	if defaults := o.loader.schema.DefaultDataValues(); defaults != nil {
+		dv, err := NewDataValues(defaults)
+		if err != nil {
+			return nil, err
+		}
+		allDvs = append(allDvs, dv)
 	}
+	for _, fileInLib := range files {
+		docs, err := o.templateFile(fileInLib)
+		if err != nil {
+			return nil, fmt.Errorf("Templating file '%s': %s", fileInLib.File.RelativePath(), err)
+		}
+		for _, doc := range docs {
+			dv, err := NewDataValues(doc)
+			if err != nil {
+				return nil, err
+			}
+			allDvs = append(allDvs, dv)
+		}
+	}
+	allDvs = append(allDvs, o.valuesOverlays...)
+	return allDvs, nil
+}
 
-	return dv, libraryValues, nil
+func (o DataValuesPreProcessing) typeAndCheck(dataValuesDoc *yamlmeta.Document) (chk yamlmeta.TypeCheck) {
```

Comment:
we typically discourage use of named returns. dont think it's necessary in this case.